test/distributed/tensor/test_redistribute.py

```diff
@@ -695,5 +695,44 @@ def test_redistribute_shard_dim_multi_dim_mesh(self):
             self.assertEqual(local_out_dt, local_expected_dt)
 
 
+class OrderedShardingTest(DTensorTestBase):
+    @property
+    def world_size(self) -> int:
+        return 4
+
+    @with_comms
+    def test_order_shard_specification(self):
+        # 2d mesh with dp and tp dim
+        mesh_row = 2
+        mesh_col = self.world_size // mesh_row
+        mesh = init_device_mesh(
+            self.device_type, (mesh_row, mesh_col), mesh_dim_names=("dp", "tp")
+        )
+
+        shard_spec_dp_first = [Shard(0, 0), Shard(0, 1)]  # shard on dp axis first
```

Comment:
First impression- this notation makes placement more confusing. Compared to Jax I think looking at dtensor placement is already confusing because it is easy to forget the names of the mesh dim names when looking at an ordered list of placements. This change adds more visual clutter by throwing more numbers in front of me.

Not sure if I have a better idea. One thought is to make 'order' be a kwonly arg; force people to write Shard(0, order=0). Is that actually better though..??

Reply:
I agree that this make placement UX more confusing, and I think it's the wrong place to encode the device ordering. I'll write down some concrete proposal